sphinx_gallery/recommender.py

```diff
@@ -0,0 +1,315 @@
+                    indices.append(j)
+            indptr.append(len(indices))
+        X = sparse.csr_matrix(
+            (data, indices, indptr), shape=(len(indptr) - 1, len(feature_names))
+        )
+        return X
+
+    @staticmethod
+    def tfidf_transformer(X):
+        """Transform a term frequency matrix into a term frequency-inverse document
+        frequency (TF-IDF) matrix.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A term frequency matrix.
+
+        Returns
+        -------
+        X_tfidf : {ndarray, sparse matrix} of shape (n_samples, n_features)
+            A tf-idf matrix of the same shape as X.
+        """
+        FLOAT_DTYPES = (np.float64, np.float32, np.float16)
+
+        if not sparse.issparse(X):
+            X = sparse.csr_matrix(X, dtype=np.float64)
+        dtype = X.dtype if X.dtype in FLOAT_DTYPES else np.float64
+
+        n_samples, n_features = X.shape
+
+        # Count the number of non-zero values for each feature in sparse X
+        if sparse.isspmatrix_csr(X):
+            df = np.bincount(X.indices, minlength=n_features)
+        else:
+            df = np.diff(X.indptr)
+        df = df.astype(dtype, copy=False)
+        # perform idf smoothing
+        df += 1
+        n_samples += 1
+        idf = np.log(n_samples / df) + 1
+
+        idf_diag = sparse.diags(
+            idf,
+            offsets=0,
+            shape=(n_features, n_features),
+            format="csr",
+            dtype=dtype,
+        )
+        X = X * idf_diag
+        X_normalized = (X.T / norm(X, axis=1)).T
+        X_normalized = sparse.csr_matrix(X_normalized, dtype=np.float64)
+
+        return X_normalized
+
+    @staticmethod
+    def cosine_similarity(X, Y=None, dense_output=True):
+        """
+        Compute the cosine similarity between two vectors X and Y.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features), default=None
+            Input data. If `None`, the output will be the pairwise
+            similarities between all samples in `X`.
+
+        dense_output : bool, default=True
+            Whether to return dense output even when the input is sparse. If
+            `False`, the output is sparse if both input arrays are sparse.
+
+        Returns
+        -------
+        cosine_similarity : ndarray of shape (n_samples_X, n_samples_Y)
+            Cosine similarity matrix.
+        """
+
+        if Y is X or Y is None:
+            Y = X
+
+        X_normalized = X / norm(X)
+        if X is Y:
+            Y_normalized = X_normalized
+        else:
+            Y_normalized = Y / norm(Y)
+
+        X_normalized = sparse.csr_matrix(X_normalized, dtype=np.float64)
+        similarity = X_normalized @ Y_normalized.T
```

Comment:
Does it makes sense to transform into sparse here if we return a dense matrix.
We could make the dot product on the dense matrix directly.